dds/src/Conversion.cpp

```diff
@@ -30,144 +29,744 @@ using namespace eprosima::fastrtps;
-                    throw DDSMiddlewareException(ss.str());
 
-           }
-       });
+    const ::xtypes::StructType& type = static_cast<const ::xtypes::StructType&>(input.type());
 
+    for (const ::xtypes::Member& member : type.members())
+    {
+        MemberId id = output->get_member_id_by_name(member.name());
+        switch (member.type().kind())
+        {
+            case ::xtypes::TypeKind::BOOLEAN_TYPE:
+            case ::xtypes::TypeKind::CHAR_8_TYPE:
+            case ::xtypes::TypeKind::CHAR_16_TYPE:
+            case ::xtypes::TypeKind::UINT_8_TYPE:
+            case ::xtypes::TypeKind::INT_8_TYPE:
+            case ::xtypes::TypeKind::INT_16_TYPE:
+            case ::xtypes::TypeKind::UINT_16_TYPE:
+            case ::xtypes::TypeKind::INT_32_TYPE:
+            case ::xtypes::TypeKind::UINT_32_TYPE:
+            case ::xtypes::TypeKind::INT_64_TYPE:
+            case ::xtypes::TypeKind::UINT_64_TYPE:
+            case ::xtypes::TypeKind::FLOAT_32_TYPE:
+            case ::xtypes::TypeKind::FLOAT_64_TYPE:
+            case ::xtypes::TypeKind::FLOAT_128_TYPE:
+            case ::xtypes::TypeKind::STRING_TYPE:
+            case ::xtypes::TypeKind::WSTRING_TYPE:
+            {
+                set_primitive_data(input[member.name()], output, id);
+                break;
+            }
+            case ::xtypes::TypeKind::ARRAY_TYPE:
+            {
+                DynamicData* array_data = output->loan_value(id);
+                set_array_data(input[member.name()], array_data, std::vector<uint32_t>());
+                output->return_loaned_value(array_data);
+                break;
+            }
+            case ::xtypes::TypeKind::SEQUENCE_TYPE:
+            {
+                DynamicData* seq_data = output->loan_value(id);
+                set_sequence_data(input[member.name()], seq_data);
+                output->return_loaned_value(seq_data);
+                break;
+            }
+            case ::xtypes::TypeKind::STRUCTURE_TYPE:
+            {
+                DynamicData* st_data = output->loan_value(id);
+                soss_to_dds(input[member.name()], st_data);
+                output->return_loaned_value(st_data);
+                break;
+            }
+            default:
+                ss << "Unsupported type: " << member.type().name();
+                throw DDSMiddlewareException(ss.str());
+
+        }
+    }
     return true;
 }
 
+void Conversion::set_sequence_data(
+        const DynamicData* c_from,
+        ::xtypes::WritableDynamicDataRef to)
+{
+    const ::xtypes::SequenceType& type = static_cast<const ::xtypes::SequenceType&>(to.type());
+    DynamicData* from = const_cast<DynamicData*>(c_from);
+
+    for (uint32_t idx = 0; idx < c_from->get_item_count(); ++idx)
+    {
+        MemberId id = idx;
+        ResponseCode ret = ResponseCode::RETCODE_ERROR;
+
+        switch (type.content_type().kind())
+        {
+            case ::xtypes::TypeKind::BOOLEAN_TYPE:
+            {
```

Comment:
Could be a macro?

Reply:
I think we don't earn anything, as the getter methods from `dds` don't follow a naming to allow a good macro here.

Reply:
ok